### PR TITLE
Fix RevocationMode and RevocationFlag errors on macOS

### PR DESCRIPTION
**Pull request by bartonjs**
For certificates which have CRL, but not OCSP, the CRL doesn't seem to be
checked, then Apple is silent about the matter.  By setting
kSecRevocationRequirePositiveResponse we get notified "RevocationResponseRequired".
That flag should be mapped to X509ChainStatusFlags.RevocationStatusUnknown.

When doing NoCheck, the OS was opportunistically checking revocation,
possibly from a cache.  So make sure that doesn't show up.

Also, filter down the results based on the RevocationFlag value.

### CODE REVIEW

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs [hunk 1]
```diff
@@ -210,10 +219,13 @@ private SafeCreateHandle PrepareCertsArray(ICertificatePal cert, X509Certificate
                     currentValue.Item2 | (int)X509ChainStatusFlags.NotValidForUsage);
             }
 
+            FixupRevocationStatus(elements, revocationFlag);
```
Comment (ianhays) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs]: Is this call still necessary if `revocationFlag == X509RevocationMode.NoCheck`?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs [hunk 2]
```diff
@@ -302,7 +314,47 @@ private void BuildAndSetProperties(Tuple<X509Certificate2, int>[] elementTuples)
...
+                // In Windows the EndCertificateOnly flag (CERT_CHAIN_REVOCATION_CHECK_END_CERT) will apply
+                // to a root if that's the only element, so we'll do the same.
+                // Start at element 1, and move to the end.
+                for (int i = 1; i < elements.Length; i++)
+                {
+                    Tuple<X509Certificate2, int> element = elements[i];
+                    X509ChainStatusFlags statusFlags = (X509ChainStatusFlags)element.Item2;
+
+                    statusFlags &= ~RevocationRelevantFlags;
+                    elements[i] = Tuple.Create(element.Item1, (int)statusFlags);
```
Comment (ianhays) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs]: So element 0 is the end certificate and the element at length-1 is the root? Then for EndCertificateOnly we remove the rev flags for all elements past the first and for ExcludeRoot we remove the rev flags for only the last element.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs (hunk 1 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs]: Technically no, but I don't think that it's something to optimize over.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs (hunk 2 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs]: Element 0 is equal to the certificate that was passed to X509Chain (which is usually and end certificate, but could be an intermediate or a root).  The last element is the root-most cert we found, which is either actually the root, or the just the last one that was found (which is why the other check looks for a "not a root" marker)

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs (hunk 1 above)
Comment (ianhays) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs]: :+1:

**Pull request merged**